src/Compilers/CSharp/Test/Syntax/Parsing/ParserErrorMessageTests.cs

```diff
@@ -2402,19 +2416,10 @@ static void M()
         D d = delegate (this object o) { };
     }
 }";
-            ParseAndValidate(test,
-                // (6,25): error CS1026: ) expected
-                Diagnostic(ErrorCode.ERR_CloseParenExpected, "this").WithLocation(6, 25),
-                // (6,25): error CS1514: { expected
-                Diagnostic(ErrorCode.ERR_LbraceExpected, "this").WithLocation(6, 25),
-                // (6,25): error CS1002: ; expected
-                Diagnostic(ErrorCode.ERR_SemicolonExpected, "this").WithLocation(6, 25),
-                // (6,30): error CS1002: ; expected
-                Diagnostic(ErrorCode.ERR_SemicolonExpected, "object").WithLocation(6, 30),
-                // (6,38): error CS1002: ; expected
-                Diagnostic(ErrorCode.ERR_SemicolonExpected, ")").WithLocation(6, 38),
-                // (6,38): error CS1513: } expected
-                Diagnostic(ErrorCode.ERR_RbraceExpected, ")").WithLocation(6, 38));
+            CreateCompilationWithMscorlib(test).VerifyDiagnostics(
+                // (6,25): error CS0027: Keyword 'this' is not available in the current context
+                //         D d = delegate (this object o) { };
+                Diagnostic(ErrorCode.ERR_ThisInBadContext, "this").WithLocation(6, 25));
```

Comment:
Agreed.  Tihs is the problem with having hte parser be overly restrictive.  All it means is htat when people write totally reasonable code, that we end up with a terrible tree.  We want the parser to avoid contextual parsing as much as possible.